tests/user_examples/extra_src/for_2D_build/k-epsilon_turbulent_model.h

```diff
@@ -0,0 +1,458 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may  *
+ * not use this file except in compliance with the License. You may obtain a*
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.       *
+ *                                                                          *
+ * ------------------------------------------------------------------------*/
+/**
+ * @file 	k-epsilon_turbulent_model.h
+ * @brief 	
+ * @details     
+ * @author Xiangyu Hu
+ */
+
+#ifndef K_EPSILON_TURBULENT_MODEL_H
+#define K_EPSILON_TURBULENT_MODEL_H
+
+#include "sphinxsys.h"
+#include <mutex>
+
+namespace SPH
+{
+namespace fluid_dynamics
+{
+	class BaseTurbuClosureCoeff
+	{
+	public:
+		explicit BaseTurbuClosureCoeff();
+		virtual ~BaseTurbuClosureCoeff() {};
+
+	protected:
+		Real Karman;
+		Real turbu_const_E;
+		Real C_mu;
+		Real TurbulentIntensity;
+		//** Closure coefficients for K *
+		Real sigma_k;
+
+		//** Closure coefficients for Epsilon *
+		Real C_l, C_2;
+		Real sigma_E;
+	};
+//=================================================================================================//
+	template <typename... T>
+	class BaseTurtbulentModel;
+	
+	template <class DataDelegationType>
+	class BaseTurtbulentModel<Base, DataDelegationType>
+		: public LocalDynamics, public DataDelegationType, public BaseTurbuClosureCoeff
+	{
+	public:
+		template <class BaseRelationType>
+		explicit BaseTurtbulentModel(BaseRelationType& base_relation);
+		virtual ~BaseTurtbulentModel() {};
+	protected:
+		StdLargeVec<Real> turbu_mu_;
+		StdLargeVec<Real> turbu_k_;
+		StdLargeVec<Real> turbu_epsilon_;
+		Real smoothing_length_;
+		Real particle_spacing_min_;
+		Real mu_;
+		StdLargeVec<Real>& rho_;
+		StdLargeVec<Vecd>& vel_;
+		int dimension_;
+	};
+//=================================================================================================//
+	template <typename... InteractionTypes>
+	class GetVelocityGradient;
+
+	template <class DataDelegationType>
+	class GetVelocityGradient<Base, DataDelegationType>
+		: public LocalDynamics, public DataDelegationType
+	{
+	public:
+		template <class BaseRelationType>
+		explicit GetVelocityGradient(BaseRelationType& base_relation);
+		virtual ~GetVelocityGradient() {};
+	protected:
+		StdLargeVec<Vecd>& vel_, & pos_;
+		StdLargeVec<Matd>& velocity_gradient_;
+		StdLargeVec<int>& is_near_wall_P1_; //** This is used to specially treat near wall region  *
+
+		//**For test*
+		StdLargeVec<Matd> velocity_gradient_wall;
+	};
+	//** Inner part *
+	template <>
+	class GetVelocityGradient<Inner<>> : public GetVelocityGradient<Base, FluidDataInner>
+	{
+	public:
+		explicit GetVelocityGradient(BaseInnerRelation& inner_relation)
+			: GetVelocityGradient<Base, FluidDataInner>(inner_relation) {};
+		virtual ~GetVelocityGradient() {};
+		void interaction(size_t index_i, Real dt = 0.0);
+	};
+	using GetVelocityGradientInner = GetVelocityGradient<Inner<>>;
```

Comment:
Is possible to write the gradient in a general way? For example, WeakGradientComputation & StrongGradientComputatoin drived from BaseGraidentComputation and velocity gradient is computed as 
StrongGradientComputation<Vecd> velocity_gradient("Velocity", complex_contact)

Reply:
OK, I will try to write a common one,